pkg/internal/metrics/metrics.go

```diff
@@ -0,0 +1,198 @@
+	// cache is ready.
+	tempCtx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
+	defer cancel()
+	return cache.WaitForCacheSync(tempCtx)
+}
+
+func collectCRsApproved(ctx context.Context, log logr.Logger, c cache.Cache, ch chan<- prometheus.Metric) {
+	list := &cmapi.CertificateRequestList{}
+	err := c.List(ctx, list)
+	if err != nil {
+		log.Error(err, "unable to list CertificateRequests")
+		return
+	}
+
+	type label struct {
+		namespace, approvedStatus string
+	}
+
+	// Let's remember the order of the labels so that we can send the metrics
+	// deterministically. Undeterministic outputs are a pain to test and debug.
+	var labels []label
+	count := make(map[label]int)
+
+	for _, cr := range list.Items {
+		approvedStatus, _ := getReason(cr.Status.Conditions, "Approved", func(c cmapi.CertificateRequestCondition) (typ, status, reason string) {
+			return string(c.Type), string(c.Status), string(c.Reason)
+		})
+		if approvedStatus == "Unknown" {
+			continue
+		}
+
+		k := label{
+			namespace:      cr.Namespace,
+			approvedStatus: string(approvedStatus),
+		}
+
+		_, exists := count[k]
+		if !exists {
+			labels = append(labels, k)
+		}
+
+		count[k] += 1
+	}
+
+	for _, key := range labels {
+		ch <- prometheus.MustNewConstMetric(
+			approvedCount,
+			prometheus.GaugeValue,
+			float64(count[key]),
+			key.namespace,
+			key.approvedStatus,
+		)
+	}
+}
+
+func collectCRsOrphans(logger logr.Logger, c cache.Cache, ch chan<- prometheus.Metric) {
+	list := &cmapi.CertificateRequestList{}
+	err := c.List(context.Background(), list)
+	if err != nil {
+		logger.Error(err, "unable to list CertificateRequests")
+		return
+	}
+
+	for _, cr := range list.Items {
+		approvedStatus, _ := getReason(cr.Status.Conditions, "Approved", func(c cmapi.CertificateRequestCondition) (typ, status, reason string) {
+			return string(c.Type), string(c.Status), string(c.Reason)
+		})
+
+		if approvedStatus != "Unknown" {
+			continue
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			orphanCount,
+			prometheus.GaugeValue,
+			1.,
+			cr.Name,
+			cr.Namespace,
+		)
+	}
+}
+
+// getReason returns the status and reason of a condition of a given type (e.g.,
+// "Ready"). When this condition type isn't found in the list of conditions,
+// getReason returns "Unknown" and "Unknown".
+//
+// Imagine you have the following conditions:
+//
+//	conditions:
+//	- type: Ready
+//	  status: "True"
+//	  reason: Issued
+//	  message: The Certificate has been issued
+//
+// You can use getReason to fetch the status and reason of that "Ready"
+// condition:
+//
+//	status, reason := getReason(conditions, "Ready", func(cond ConnectionCondition) (typ, sta, reason string) {
+//	  return string(cond.Type), string(cond.Status), string(cond.Reason)
+//	})
+func getReason[C any](conditions []C, typ string, readCond func(cond C) (typ, status, reason string)) (status, reason string) {
+	for _, cond := range conditions {
```

Comment:
Maybe use this utility function instead:

https://github.com/kubernetes/apimachinery/blob/a1b8da8152f9991a6691ae94bd6b6e3c34a590bd/pkg/api/meta/conditions.go#L91

Reply:
Unfortunately, we use a custom type `cmapi.CertificateRequestCondition` instead of `metav1.Condition` 😞